lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+  if (anyFailed)
+    return failure();
+
+  // If we have found no annotations, there is nothing to do. We just leave this
+  // module unannotated, which will cause it to inherit a reset domain from its
+  // instantiation sites.
+  if (!ignore && !reset) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "No reset annotation for " << module.getName() << "\n");
+    return success();
+  }
+
+  // If we have found multiple annotations, emit an error and abort.
+  if (conflictingAnnos.size() > 1) {
+    auto diag = module.emitError("multiple reset annotations on module '")
+                << module.getName() << "'";
+    for (auto &annoAndLoc : conflictingAnnos)
+      diag.attachNote(annoAndLoc.second)
+          << "Conflicting " << annoAndLoc.first.getClassAttr() << ":";
+    return failure();
+  }
+
+  // Dump some information in debug builds.
+  LLVM_DEBUG({
+    llvm::dbgs() << "Annotated reset for " << module.getName() << ": ";
+    if (ignore)
+      llvm::dbgs() << "no domain\n";
+    else if (auto arg = reset.dyn_cast<BlockArgument>())
+      llvm::dbgs() << "port " << getModulePortName(module, arg.getArgNumber())
+                   << "\n";
+    else
+      llvm::dbgs() << "wire "
+                   << reset.getDefiningOp()->getAttrOfType<StringAttr>("name")
+                   << "\n";
+  });
+
+  // Store the annotated reset for this module.
+  assert(ignore || reset);
+  annotatedResets.insert({module, reset});
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Domain Construction
+//===----------------------------------------------------------------------===//
+
+/// Gather the reset domains present in a circuit. This traverses the instance
+/// hierarchy of the design, making instances either live in a new reset domain
+/// if so annotated, or inherit their parent's domain. This can go wrong in some
+/// cases, mainly when a module is instantiated multiple times within different
+/// reset domains.
+LogicalResult InferResetsPass::buildDomains(CircuitOp circuit) {
+  // Gather the domains.
+  auto instGraph = getAnalysis<InstanceGraph>();
+  auto module = dyn_cast<FModuleOp>(instGraph.getTopLevelNode()->getModule());
+  if (!module) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Skipping circuit because main module is no `firrtl.module`");
+    return success();
+  }
+  buildDomains(module, InstancePath{}, Value{}, instGraph);
+
+  // Report any domain conflicts among the modules.
+  bool anyFailed = false;
+  for (auto &it : domains) {
+    auto module = cast<FModuleOp>(it.first);
+    auto &domainConflicts = it.second;
+    if (domainConflicts.size() <= 1)
+      continue;
+
+    anyFailed = true;
+    SmallDenseSet<Value> printedDomainResets;
+    auto diag = module.emitError("module '")
+                << module.getName()
+                << "' instantiated in different reset domains";
+    for (auto &it : domainConflicts) {
+      auto &domain = it.first;
```

Comment:
Please don't use 'auto' in cases like this where it isn't obvious what it.first is.